pkg/api/api.go

```diff
@@ -0,0 +1,233 @@
+
+// unique keys for context setting
+type key int
+
+const (
+	apiKeyCtxKey key = iota
+)
+
+// errors
+const (
+	errContextCanceled = "Context was canceled."
+)
+
+// Option allows to set optional parameters in the server
+type Option func(*Server)
+
+// WithKeys registers API authentication keys with the server
+func WithKeys(keys []string) Option {
+	return func(s *Server) {
+		for _, key := range keys {
+			s.keys[key] = struct{}{}
+		}
+	}
+}
+
+// WithLogger provides the api with access to the program level logger. It is recommended to use this option
+func WithLogger(l log.Logger) Option {
+	return func(s *Server) {
+		s.logger = l
+	}
+}
+
+// WithMetricsExport switches on metrics export
+func WithMetricsExport() Option {
+	return func(s *Server) {
+		s.metrics = true
+	}
+}
+
+// WithRateLimits applies custom request rate limits to the API. Setting any of the values to zero deactivates the rate limiting for that particular part.
+func WithRateLimits(perMin, burst, concurrentRequests int) Option {
+	return func(s *Server) {
+		s.perMinRateLimit = perMin
+		s.burstLimit = burst
+		s.concurrentAccessLimit = concurrentRequests
+	}
+}
+
+func getLoggerHandler(logger log.Logger) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			logger.Debugf("%s %s", r.Method, r.URL.Path)
+			next.ServeHTTP(w, r)
+		}
+		return http.HandlerFunc(fn)
+	}
+}
+
+func getMetrics(w http.ResponseWriter, r *http.Request) {
+	expvar.Handler().ServeHTTP(w, r)
+}
+
+// New creates a base router for goProbe's APIs and provides the metrics export out of the box
+func New(host, port string, manager *capture.Manager, opts ...Option) (*Server, error) {
+
+	s := &Server{
+		root:                  "/",
+		keys:                  make(map[string]struct{}),
+		perMinRateLimit:       defaultPerMinRateLimit,
+		burstLimit:            defaultBurstLimit,
+		concurrentAccessLimit: defaultConcurrentAccessLimit,
+		logger:                log.NewDevNullLogger(), // default is logging to nothing
+	}
+
+	if host == "" {
+		host = "localhost"
+	}
+	if port == "" {
+		return nil, fmt.Errorf("no port provided")
+	}
+	s.location = host + ":" + port
+
+	// apply options
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	// initialize currently supported APIs
+	s.apis = append(s.apis, v1.New(manager, v1.WithLogger(s.logger)))
+
+	r := chi.NewRouter()
+
+	// setup a good base middleware stack
+	r.Use(middleware.RequestID)
+	r.Use(middleware.RealIP)
+
+	// only use the logging middleware if a logger was specifically provided
+	switch s.logger.(type) {
+	case *log.DevNullLogger:
+	default:
+		r.Use(reqLogger) // prints to stdout at the moment
+	}
+	r.Use(middleware.Recoverer)
+
+	// Set a timeout value on the request context (ctx), that will signal
+	// through ctx.Done() that the request has timed out and further
+	// processing should be stopped.
+	r.Use(middleware.Timeout(30 * time.Second))
```

Comment:
I guess that we could easily create requests that last longer than 30s for big databases, right (at least if and once we allow querying results)? Maybe this could be made another functional option to allow control over the behavior?